### ISSUE TITLE
[Smallfix] Lambda Pattern Match

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#268]
Consider the following code:

```
def f(o: Option[(Int, Int]): Option[Int] = option/map((x, y) -> x + y), o)
```

Unfortunately this code does not work. 

The problem is that the lambda has type: `(Int, Int) -> Int)` but you actually need it to have type `((Int, Int)) -> Int`. However, to do that, you have to write:

```
def f(o: Option[(Int, Int]): Option[Int] = option/map(p -> let (x, y) = p; x + y), o)
```

to destruct the pair.

We could try to fix this in several ways:

1. We could introduce (a lot of ) compiler magic to try and make the types work. I.e. when unification fails, we would try to unwrap/wrap tuples to make them fit.
2. We could change the definition of lambdas to allow (limited) pattern matching in the parameters. So, when you write `(x, y) -> ...` you mean a lambda with two arguments, but when you write `((x, y)) -> ...` you mean a function with one argument (which is a pair) and you want it unpacked into `x` and `y`.

I think 2. is doable, but it does require to programmer to (1) understand when there is a tupling issue, and (2) to put the extra parens where necessary.

Any other ideas? Is there some well-known solution to this issue?

**Comment by mhyee** [comment]
I don't really like 1. because it's implicit magic and not obvious at all to the programmer. 2. is a bit better, but there's still a bit of magic. Also, it raises the question about what pattern matching to allow or disallow.

Actually, doesn't Scala get around that with partial functions and pattern matching? You would write `case (x, y) => x + y` as the definition of a partial function (which does pattern matching), instead of the lambda function `(x, y) => x + y`.

That's probably the closest thing to our solution 2.

I suppose another approach is some library function, but it's probably too messy without currying. (Sorry if my generic syntax is wrong.)

```
def unwrap2[a,b,c](f: (a, b) -> c, p: (a, b)): c =
  let (x, y) = p;
    f(x, y)

def f(o: Option[(Int, Int]): Option[Int] = option/map(p -> unwrap2((x, y) -> x + y, p), o)
```

Yeah, after writing that out, it's pretty ugly. I suppose it works better in a language like Haskell when you don't have parens and everything is curried.

Also, does OCaml have something similar? I only played around it for a bit, and it was a few years ago, but I vaguely remember some confusion about a function that takes a pair versus a function that takes two (non-tuple) arguments.

**Comment by magnus-madsen** [comment]
If the grammar can allow it, how about changing the syntax of lambdas to `pat => body`?

Then you could even do:
```
enum Shape {
  Rect(Int, Int)
}

def f(xs: List[Shape): Int = List/map(Rect(width, height) -> width*height)
```

(Obviously the pattern match could fail...)

**Comment by magnus-madsen** [comment]
I want to improve this code:

```
            match redex(e1) with {
                case None => None
                case Some((e, ec)) =>
                    // The first argument has a redex `e` in evaluation context `ec`.
                    // Extend the evaluation context with `e2`.
                    Some((e, App1(ec, e2)))
            }
```

I cannot simply change the grammar to `pat -> body`, because then what does `(x, y) -> e` then mean? 

Is it a lambda of two arguments (x, y) or a lambda of a single argument which should be a pair `(x, y)`?

A hackish solution would be to add a production `(pat) -> e` (note the extra parenthesis). Then you could write e.g. `map((x, y) -> ...)`, and you could write `map((Red(x)) -> ...)`. Notice how the latter case requires extra parenthesis. Technically, this solution would make `(x) -> ...` ambiguous (i.e. is it a regular lambda or a pattern matching lambda), but this can be overcome by changing the grammar slightly or relying on the ordered choice.

A more verbose, but perhaps clearer solution, would be to re-use the `match` keyword and add a production `match pat -> e`. For example, the above code would become:

```
Option/map(match (e, ec) -> (e, App1(ec, e2)), redex(e1))
```

**Comment by mhyee** [comment]
(That code snippet is basicaly the textbook example of why monads are useful, but that's probably more than what we need in Flix right now.)

It looks like #284 is the PR for this issue (it didn't auto-link or auto-reference because you need to put the `#284` in the body of the PR and not the title).

From skimming the code, it looks like the PR is still in progress, and you're implementing `match pat -> e`? I like that proposal a lot more than `(pat) -> e`.

**Comment by magnus-madsen** [comment]
Yeah, I don't see a clean solution without putting a keyword in front. You just cannot distinguish between a multiple argument lambda and a pattern matching lambda on a single tuple. At least I don't see it now.

Certainly the match-lambda does not break anything and could easily be refactored if we change our minds. One issue is that perhaps the word `case` would be more appropriate (like Scala), but I sort of prefer `match` myself.

**Comment by magnus-madsen** [comment]
Merged in #284 .
